program-analysis/echidna/Exercise-X.md

```diff
@@ -0,0 +1,440 @@
+  
+  According to the ERC20 spec, the `approve` function should not fail if the caller has enough tokens to make the approval.
+  
+   ```
+   function approve_should_not_fail_if_caller_has_enough_tokens(uint256 amount)
+        public
+    {
+        // pre-conditions
+        uint256 playerBalance = naughtCoin.balanceOf(player);
+        if (playerBalance >= amount) {
+            // actions
+            bool success1 = naughtCoin.approve(bob, amount);
+            // post-conditions
+            assert(success1);
+        }
+    }
+   ```
+
+</details>
+
+<details>
+  <summary>5th property</summary>
+  
+  Player should not `burn` tokens before the `timeLock` period.
+  What is token burning?
+  
+  You can burn tokens by sending them to the `0` address or by sending them to a non existent address. Sending to the `0` address is not possible because of the `0` address checks in the `ERC20` OZ standard. Sending them to any address shouldn't be possible as well because of the `timeLock`.
+  
+  This property would be invalidated if any of our previous property were invalidated (which seems to not be the case at the moment). Let's leave this property for now and move on.
+
+</details>
+
+<details>
+  <summary>6th property</summary>
+  
+  The token transfer via `transferFrom` should fail if the current `block.timestamp < timelock` and/or spender has enough allowance.
+  
+  This property in its base form is the same as the one with `transfer` function.
+  
+   ```
+   function transfer_from_should_fail_before_timelock_period(uint256 amount)
+        public
+    {
+        // pre-conditions
+        uint256 playerBalanceBefore = naughtCoin.balanceOf(player);
+        uint256 bobBalanceBefore = naughtCoin.balanceOf(bob);
+        uint256 currentTime = block.timestamp;
+        if (currentTime < naughtCoin.timeLock()) {
+            // actions
+            bool success1 = naughtCoin.transferFrom(player, bob, amount);
+            if (success1) {
+                emit AssertionFailed(amount);
+            }
+            // post-conditions
+            assert(
+                naughtCoin.balanceOf(player) == playerBalanceBefore &&
+                    naughtCoin.balanceOf(bob) == bobBalanceBefore
+            );
+        }
+    }
+   ```
+
+Run Echidna and see what you get. Echidna emits the `AssertionFailed` event with the `amount` of `0`. It means that the `transferFrom` succeeded. If you look at the corpus, you will see that our post-conditions were invalidated. Echidna found a way to change player's balance.
+
+The property `playerBalance == InitialSupply` still holds. You need to guide Echidna a little bit more for it to invalidate this property.
+
+Try to get into the mindset of exploring the system you are auditing. Since, we have discovered something new about the system, a `transferFrom` function does not fail, as we initially expected, the next logical step would be to test this functionality and make sure that it works properly.
```

Comment:
```suggestion
Try to get into the mindset of exploring the system you are auditing. Since we have discovered something new about the system, a `transferFrom` function does not fail, as we initially expected, the next logical step would be to test this functionality and ensure that it works properly.
```

Reply:
Change wording